### PR TITLE
gas miners scale with power MK2

### PR DESCRIPTION
**Pull request by pedr69**
image

## What this does
- [x] makes gas miners able to draw ~excess~more power from the APC to mine faster
- [x] adds a right click verb to the miner to change the requested power
~- [x] connects the miner to the grid when wrenched on top of a wire node~
~- [ ] connects the miner to the grid when a wire is built under it~
~- [ ] shows the miner's power draw on the power monitor~
~- [ ] shows the miner's power draw when multitooling consistently~
~- [x] reduces the mining speed back to default if the power wire gets cut while running~
~- [x] can by abused by #33737~
- [x] scale the external pressure limit with power, instead of just disabling it
- [x] adds output gauges to let you see the mining speed
- [x] removes nonfunctional "gas giant" miner and "gas sink", aka hyperspace gas miner
- [x] allows derelict crabs to generate xboxhueg amounts of gas since they don't have an APC

## Why it's good
i dunno you tell me

## Changelog

:cl:
 * rscadd: gas miners now show their output speed
 * rscadd: gas miners can now draw additional power set via rightclick
 * rscadd: gas miners now mine faster and with a larger pressure limit the more power they consume

### CODE REVIEW

File: code/game/machinery/atmoalter/gas_mine.dm [hunk 1]
```diff
@@ -7,29 +9,81 @@
...
+		air_contents = null
+	..()
+
+
+//update gas creation speed into air_contents
+/obj/machinery/atmospherics/miner/proc/set_rate(var/internal_pressure)
+	air_contents.remove(air_contents.total_moles)//set to 0
+	//rate is in mols
+	rate = internal_pressure * air_contents.volume / (R_IDEAL_GAS_EQUATION * air_contents.temperature)
+	//this is ugly
```
Comment (adacovsk) [inline code/game/machinery/atmoalter/gas_mine.dm]: Replace this with `for(var/i = 1, i <= gases.len, i++)
		air_contents.adjust_gas(gases[i], gases[gases[i]]*rate)`

Comment (adacovsk) [reply code/game/machinery/atmoalter/gas_mine.dm]: #33720 has the better code

File: code/game/machinery/atmoalter/gas_mine.dm [hunk 2]
```diff
@@ -111,138 +163,42 @@
...
-/obj/machinery/atmospherics/miner/gas_giant/initialize()
-	..()
-	AddAir()
-
-/obj/machinery/atmospherics/miner/gas_giant/AddAir()
-	if(ticker)
-		air_contents.copy_from(gas_giant.GM)
-
-
-/obj/machinery/atmospherics/miner/mixed_nitrogen
```
Comment (adacovsk) [inline code/game/machinery/atmoalter/gas_mine.dm:209]: #33720 contains the mixed_nitrogen mixed_oxygen code

File: code/game/machinery/atmoalter/gas_mine.dm (hunk 1 above)
Comment (pedr69) [reply code/game/machinery/atmoalter/gas_mine.dm:53]: this is referring to the part that isn't quoted there, the giant blocks of adjust_multi, right?

File: code/game/machinery/atmoalter/gas_mine.dm (hunk 2 above)
Comment (pedr69) [reply code/game/machinery/atmoalter/gas_mine.dm:209]: oh right i just noticed half the file at the end is removed

File: code/game/machinery/atmoalter/gas_mine.dm (hunk 1 above)
Comment (adacovsk) [reply code/game/machinery/atmoalter/gas_mine.dm]: yes

**Comment by pedr69** [comment]
the gas sink never seems to work, since datum/zlevel/hyperspace doesn't exist??
the gas giant, similarly, doesn't seem to work and i have no idea how to view _var/global/datum/gas_giant/gas_giant_ while ingame

**Comment by adacovsk** [comment]
> the gas sink never seems to work, since datum/zlevel/hyperspace doesn't exist?? the gas giant, similarly, doesn't seem to work and i have no idea how to view _var/global/datum/gas_giant/gas_giant_ while ingame

Yeah I don't know why I couldn't get it to work either, very strange

**Comment by pedr69** [comment]
how would i make the gasminers inherit from _machinery/power_ ?
or do i need to copy half the code

**Comment by adacovsk** [comment]
> how would i make the gasminers inherit from _machinery/power_ ? or do i need to copy half the code

That's the issue. There are ways to get around it though. Some of my code already shows you how, but there's a lot I don't know

**Comment by Eneocho** [comment]
I still think that a tweakable-via-computer-console method would've worked better instead of the priority system, since you could exactly control how much juice you feeding the thing.

**Comment by adacovsk** [comment]
> I still think that a tweakable-via-computer-console method would've worked better instead of the priority system, since you could precisely control how much juice you're feeding the thing. Just put the pressure, it calculates the power required (or viceversa) and overrides the internal value.

There's still the issue from my PR that it draws from the APC rather than the powernet itself

**Comment by pedr69** [comment]
working on a direct power connection under the miner and a multitool menu to set consumption
and i realise powerdraw isn't set via multitool anywhere so it wouldn't be consistent
i've never made a GUI before...

**Comment by jellyveggie2** [comment]
I saw what I think was pedr69's asking for help yesterday in the thread a few hours ago and gave it [a reply](https://arch.b4k.co/vg/thread/407899027/#408298889), so this is expanding on that now I've realized there was a PR up.

To summarize, you want to use `/datum/power_connection`. Look up _/code/modules/power/components.dm_
There's a working example of this in the radio broadcaster (_broadcast.dm_), among others, searching for "_/datum/power_connection_" should show bring up more.

#### Adding the power component to the miner:
```
/obj/machinery/atmospherics/miner
    ...
    var/datum/power_connection/consumer/power_connection = null
    ...

/obj/machinery/atmospherics/miner/New()
    ..()
    power_connection = new(src)
    power_connection.power_priority = POWER_PRIORITY_??? // starting priority. defaults to NORMAL if not set (same priority as APCs)
    power_connection.monitoring_enabled = TRUE // This makes the miner show up on the monitor. Defaults to FALSE if not set
    ...

/obj/machinery/atmospherics/miner/Destroy()
    ...
    if(power_connection)
        qdel(power_connection)
        power_connection = null
    ..()
```

Some procs I believe will be important to you:
* `power_connection.use_power(100)`: Demands 100W from the area's APC
* `power_connection.add_load(100)`: Demands 100W from the powernet. Must be connected to a powernet first (call `connect()`)
* `power_connection.get_satisfaction()`: Tells you what percentage of the load you requested last tick got fulfilled this tick. Calling `add_load(100)` one tick and getting `get_satisfaction() == 0.75` the next tick means you're only getting 75W.
* `power_connection.connect()`: Attempts to connect to the wire node underneath. If you pass it a cable as an argument, it'll connect to that cable instead and skip checking for nodes.
* `power_connection.disconnect()`: What it says on the tin. Do not forget to call it when disconnecting from the powernet.

#### Handling power consumption
You have a var tracking how much power you *want* the miner to use (`power_load`). You also need a var along the lines of  `power_requested_old`, which needs to be reset to 0 whenever you disconnect or turn off the miner. 
So, do this inside `process()`:
1. Do `var/power_received = power_requested_old * power_connection.get_satisfaction()`.
2. Produce gas according to how much power you have received (`power_received`, not `power_load`!).
3. Call `power_connection.add_load(power_load)`.
4. Do `power_requested_old = power_load`.
5. Do `power_connection.monitor_demand = power_load`.

The reason why a `power_requested_old` is needed is that `get_satisfaction()` is giving you the satisfaction for the *previous* tick, so special care must be had on step 1 for the first iteration because the satisfaction calculations will not be including this tick's demand yet! 
Consider a case where you hook a gas miner with power_load 900W to a network producing 50W and consuming 100W. Before you hooked the miner, satisfaction was at 50%, and after hooking the miner, satisfaction should be 5%, meaning your miner would be getting a measly 45W. 
However, during your first iteration, naively doing `power_load * power_connection.get_satisfaction()` would net you 450W, and the miner would produce 10x the amount of gas it should, because as far as the network is concerned, the power draw was only 100W back then, not 1kW!

#### Handling the wire connection
For a direct connection, just do as the radio broadcaster does: Extend `wrenchAnchor()` to have the miner call `connect()` and `disconnect()` when anchored and unanchored, respectively. Do not forget to reset `power_requested_old`.

You can also do something fancy like having the miner locate the area's APC, then the cable under the APC, and call `connect(apc_cable)`. Or maybe have a external computer (per Eneocho's "_tweakable-via-computer-console_" method) wired to the powernet like the power monitor does, and have the console call something along the lines of `miner.power_connection.connect(console.cable)` when connecting a miner.

This is about all I can research for tonight, plus what I can recall from last time I talked about this with Adacovsk on Discord, pre-purge I believe.

**Comment by pedr69** [comment]
does there exist a `machinery/proc/a_power_cable_just_got_placed_under_this` for me to hook into
otherwise i'm not sure how elso to connect to a wire properly without checking each tick

**Comment by adacovsk** [comment]
> does there exist a `machinery/proc/a_power_cable_just_got_placed_under_this` for me to hook into otherwise i'm not sure how else to connect to a wire properly without checking each tick

Just check when there's a power change. Also, if you go this route, you will have to change every single map to include the wire

**Comment by Eneocho** [comment]
I dislike the wire requirement, please don't implement it wired.

**Comment by pedr69** [comment]
i've coded the apc wire connection but it doesn't feel right - it's not consistent with anything
machines either draw from the APC and its battery or from a wirenode directly under the machine
drawing from the wirenode under the APC seems silly

**Comment by pedr69** [comment]
the wires have broken me. i kneel. i yield.
they draw straight out the APC now.

**Pull request merged**